dawn/src/dawn/Optimizer/PassMultiStageMerger.cpp

```diff
@@ -0,0 +1,156 @@
+  //    G   -----+                 +
+  //    E     B  |  B     X     B  |
+  //        -----+-----------------+
+  //
+  std::vector<iir::LoopOrderKind> possibleLoopOrders;
+
+  if(multiStageLoopOrder == iir::LoopOrderKind::Parallel &&
+     stageLoopOrder == iir::LoopOrderKind::Parallel)
+    possibleLoopOrders = {iir::LoopOrderKind::Parallel, iir::LoopOrderKind::Forward,
+                          iir::LoopOrderKind::Backward};
+  else if(stageLoopOrder == iir::LoopOrderKind::Parallel)
+    possibleLoopOrders.push_back(multiStageLoopOrder);
+  else
+    possibleLoopOrders.push_back(stageLoopOrder);
+
+  if(multiStageDependencyGraph.empty())
+    return ReturnType(multiStageDependencyGraph, possibleLoopOrders.front());
+
+  // Check if the resulting graph is no longer a DAG (i.e., cycles exist)
+  if(!multiStageDependencyGraph.isDAG())
+    return ReturnType(std::nullopt, multiStageLoopOrder);
+
+  // Check all possible loop orders if there aren't any vertical conflicts
+  for(auto loopOrder : possibleLoopOrders) {
+    auto conflict = hasVerticalReadBeforeWriteConflict(multiStageDependencyGraph, loopOrder);
+    if(!conflict.CounterLoopOrderConflict)
+      return ReturnType(multiStageDependencyGraph, loopOrder);
+  }
+
+  return ReturnType(std::nullopt, multiStageLoopOrder);
+}
+
+/// @brief Check if we can merge the first multistage into the other, possibly changing the loop
+/// order.
+/// @returns The new dependency graphs of the multi-stage (or NULL) and the new loop order
+template <typename ReturnType =
+              std::pair<std::optional<iir::DependencyGraphAccesses>, iir::LoopOrderKind>>
+ReturnType isMergable(const iir::MultiStage& thisMS, const iir::MultiStage& otherMS) {
+  iir::LoopOrderKind thisLoopOrder = thisMS.getLoopOrder();
+  for(const auto& thisStage : thisMS.getChildren()) {
+    auto dependencyGraphLoopOrderPair = isMergable(*thisStage, thisLoopOrder, otherMS);
+    if(dependencyGraphLoopOrderPair.first) {
+      return dependencyGraphLoopOrderPair;
+    }
+  }
+  return ReturnType(std::nullopt, thisLoopOrder);
+}
+
+PassMultiStageMerger::PassMultiStageMerger(OptimizerContext& context)
+    : Pass(context, "PassMultiStageMerger") {
+  dependencies_.push_back("PassSetStageGraph");
+}
+
+bool PassMultiStageMerger::run(const std::shared_ptr<iir::StencilInstantiation>& instantiation) {
+  const int maxBoundaryExtent = context_.getOptions().MaxHaloPoints;
+  for(const auto& stencil : instantiation->getStencils()) {
+    if(stencil->getChildren().size() < 2)
+      continue;
+
+    auto& metadata = instantiation->getMetaData();
+    std::unique_ptr<iir::Stencil> newStencil = std::make_unique<iir::Stencil>(
+        metadata, stencil->getStencilAttributes(), stencil->getStencilID());
+    newStencil->setStageDependencyGraph(
+        iir::DependencyGraphStage(*stencil->getStageDependencyGraph()));
+
+    // NOTE: The Stages in the same multi-stage are assumed to have no counter loop-order vertical
```

Comment:
what does that mean? I can write 

```
a = 1, b = 0
vertical_region(forward){
  a = 10
  b = a[k+1]
}
a=10, b=1
```
this is compliant with our parallel model

Reply:
Removed this statement.